fbgemm_gpu/include/fbgemm_gpu/enum_utils.h

```diff
@@ -15,9 +15,11 @@ namespace fbgemm_gpu {
 
 #define FBGEMM_GPU_ENUM_CREATE_TAG(module_name)                                \
   struct fbgemm_gpu_enum_tag_##module_name {};                                 \
-  extern template enum_registration<struct fbgemm_gpu_enum_tag_##module_name>* \
+  template <> enum_registration<struct fbgemm_gpu_enum_tag_##module_name>*     \
       enum_registration<                                                       \
-          struct fbgemm_gpu_enum_tag_##module_name>::registration_list;
+          struct fbgemm_gpu_enum_tag_##module_name>::registration_list;        \
+  extern template class enum_registration<                                     \
+      struct fbgemm_gpu_enum_tag_##module_name>;
```

Comment:
Would you mind explaining why this change is needed for ROCm?

Reply:
When using hipcc, the compiler complains `error: explicit specialization of 'registration_list' after instantiation`. 
Based on https://stackoverflow.com/questions/54483114/specialization-of-a-template-class-static-member-variable-when-also-using-exter, one had to explicitly declare the static class member for the specialized template prior to instantiating the specialized template later.

Reply:
The change in this file here caused an internal issue with the error message like `stderr: ld.lld: error: undefined symbol: fbgemm_gpu::enum_registration<fbgemm_gpu::fbgemm_gpu_enum_tag_uvm>::enum_query[abi:cxx11]()`, so it wasn't merged into the trunk. Feel free to add another PR to incorporate the changes and we can do more testings.

Reply:
> The change in this file here caused an internal issue with the error message like `stderr: ld.lld: error: undefined symbol: fbgemm_gpu::enum_registration<fbgemm_gpu::fbgemm_gpu_enum_tag_uvm>::enum_query[abi:cxx11]()`, so it wasn't merged into the trunk. Feel free to add another PR to incorporate the changes and we can do more testings.

Hi @jianyuh. Is this error occurs when importing fbgemm_gpu or when compiling the project? We don't have this issue on our side. Would you please provide more info? such that we could reproduce the error. Thank you.

Reply:
@liligwu Thanks. We looked at this issue. It seems that the difference comes from NVCC vs AMD Clang. You can reproduce it by running `uvm_test.py` with CUDA enabled.

https://godbolt.org/z/johx4fEME is a very simplified current code in FBGEMM (the core code is borrowed from https://stackoverflow.com/questions/54483114/specialization-of-a-template-class-static-member-variable-when-also-using-exter). This code is acceptable for GCC (and supposedly GCC-based NVCC) while not acceptable for Clang (and supposedly Clang-based AMD compiler).

I am not sure which compiler's behavior is correct in terms of C/C++, but I'd appreciate it if you could find a solution that works for both platforms.

Reply:
> @liligwu Thanks. We looked at this issue. It seems that the difference comes from NVCC vs AMD Clang. You can reproduce it by running `uvm_test.py` with CUDA enabled.
> 
> https://godbolt.org/z/johx4fEME is a very simplified current code in FBGEMM (the core code is borrowed from https://stackoverflow.com/questions/54483114/specialization-of-a-template-class-static-member-variable-when-also-using-exter). This code is acceptable for GCC (and supposedly GCC-based NVCC) while not acceptable for Clang (and supposedly Clang-based AMD compiler).
> 
> I am not sure which compiler's behavior is correct in terms of C/C++, but I'd appreciate it if you could find a solution that works for both platforms.

We'll resolve this issue in the next PR.